dbms/src/Flash/Coprocessor/DAGExpressionAnalyzer.cpp

```diff
@@ -192,6 +200,153 @@ void appendWindowDescription(
+        return;
+
+    if (!window_desc.order_by.empty())
+    {
+        const Block & sample_block = actions->getSampleBlock();
+        const String & order_by_col_name = window_desc.order_by[0].column_name;
+        const ColumnWithTypeAndName & order_by_col_type_and_name = sample_block.getByName(order_by_col_name);
+
+        window_desc.order_by_col_type = order_by_col_type_and_name.type->getTypeId();
+        window_desc.is_desc = (window_desc.order_by[0].direction == -1);
+    }
+}
+
+// Add a function generating a new auxiliary column that help the implementation of range frame type
+std::pair<String, String> addRangeFrameAuxiliaryFunctionAction(
+    DAGExpressionAnalyzer * analyzer,
+    ExpressionActionsPtr & actions,
+    const tipb::Window & window)
+{
+    // Execute this function only when the frame type is Range
+    if (window.frame().type() != tipb::WindowFrameType::Ranges)
+        return std::make_pair("", "");
+
+    String begin_aux_col_name;
+    String end_aux_col_name;
+
+    // For begin frame
+    if (window.frame().start().has_frame_range())
+        begin_aux_col_name = DAGExpressionAnalyzerHelper::buildFunction(analyzer, window.frame().start().frame_range(), actions);
+    // For end frame
+    if (window.frame().end().has_frame_range())
+        end_aux_col_name = DAGExpressionAnalyzerHelper::buildFunction(analyzer, window.frame().end().frame_range(), actions);
+
+    return std::make_pair(begin_aux_col_name, end_aux_col_name);
+}
+
+WindowDescription createAndInitWindowDesc(DAGExpressionAnalyzer * const analyzer, const tipb::Window & window)
+{
+    WindowDescription window_description;
+    window_description.partition_by = analyzer->getWindowSortDescription(window.partition_by());
+    window_description.order_by = analyzer->getWindowSortDescription(window.order_by());
+    if (window.has_frame())
+    {
+        window_description.setWindowFrame(window.frame());
+    }
+
+    return window_description;
+}
+
+void initBeforeWindow(
+    DAGExpressionAnalyzer * analyzer,
+    WindowDescription & window_desc,
+    ExpressionActionsChain & chain,
+    const tipb::Window & window,
+    ExpressionActionsChain::Step & step)
+{
+    // Prepare auxiliary function for range frame type
+    auto aux_col_names = addRangeFrameAuxiliaryFunctionAction(analyzer, step.actions, window);
+
+    analyzer->appendWindowColumns(window_desc, window, step.actions);
+    // set required output for window funcs's arguments.
+    for (const auto & window_function_description : window_desc.window_functions_descriptions)
+    {
+        for (const auto & argument_name : window_function_description.argument_names)
+            step.required_output.push_back(argument_name);
+    }
+
+    window_desc.before_window = chain.getLastActions();
+    setAuxiliaryColumnInfo(step.actions, window_desc, aux_col_names.first, aux_col_names.second, window);
+
+    // We should not call finalize here as some useful columns will be unexpected removed
```

Comment:
The handling here is a bit weird.
why not setting require_output?

Reply:
> The handling here is a bit weird. why not setting require_output?

Because these auxiliary column don't appear in sample_block, and it will raise error.

Reply:
https://github.com/pingcap/tiflash/pull/7629/files#r1270426007
But here getting aux col from sample block.

Reply:
> https://github.com/pingcap/tiflash/pull/7629/files#r1270426007 But here getting aux col from sample block.

done